prime-dotnet-webapi/Services/ReportingService.cs

```diff
@@ -0,0 +1,125 @@
+
+        public ReportingService(
+            ApiDbContext context,
+            ILogger<ReportingService> logger,
+            ICollegeLicenceClient collegeLicenceClient)
+            : base(context, logger)
+        {
+            _collegeLicenceClient = collegeLicenceClient;
+        }
+
+        public async Task<string> PopulatePractitionerTableAsync(DateTime? startDate = null, DateTime? endDate = null)
+        {
+            var result = "";
+            try
+            {
+                if (startDate == null)
+                {
+                    startDate = DateTime.Now.AddDays(-CalculationPeriodInDays);
+                }
+
+                if (endDate == null)
+                {
+                    endDate = DateTime.Now;
+                }
+
+                // get all approved enrollee
+                var enrolleeLicences = _context.Enrollees
+                    .Where(e => e.GPID != null && e.Certifications.Any(c => c.Prefix != null))
+                    .Select(e => new
+                    {
+                        e.Certifications.FirstOrDefault().CollegeCode,
+                        e.Certifications.FirstOrDefault().PractitionerId,
+                        e.Certifications.FirstOrDefault().LicenseNumber,
+                        // do not pull prefix from LicenseDetail since we are not sure if prescribing or not
+                        // and the Prefix here has been verified from PharmaNet API
+                        e.Certifications.FirstOrDefault().Prefix
+                    });
+                // query the unauthorized access practitioner ID from pharmanet transaction log table
+                var questionablePractitionerIds = await _context.PharmanetTransactionLogs
+                    .Where(l => l.TxDateTime >= startDate && l.TxDateTime <= endDate)
+                    .Where(l => !enrolleeLicences.Where(e =>
+                        // for college BCCNM (code 3), compare PharmaNet ID of the college license to Practitioner Id of the log
+                        (e.CollegeCode == 3 && e.PractitionerId == l.PractitionerId && e.Prefix == l.CollegePrefix)
+                        //for other college, use License Number
+                        || (e.CollegeCode != 3 && e.LicenseNumber == l.PractitionerId && e.Prefix == l.CollegePrefix)).Any())
+                    .Where(l => !_context.Practitioner.Where(p => p.PracRefId == l.CollegePrefix && p.CollegeId == l.PractitionerId).Any())
+                    .Select(l => new
+                    {
+                        l.PractitionerId,
+                        l.CollegePrefix
+                    }).ToListAsync();
+
+                foreach (var p in questionablePractitionerIds)
+                {
+                    _context.Practitioner.Add(new Practitioner
+                    {
+                        CollegeId = p.PractitionerId,
+                        PracRefId = p.CollegePrefix
+                    });
+                }
```

Comment:
If there are multiple records in `PharmanetTransactionLog` for the same `PractitionerId`/`CollegePrefix` pair, do we want multiple `Practitioner` records created?

Reply:
Good catch. it is fixed.